tests/spread/general/entrypoint/rockcraft.yaml

```diff
@@ -0,0 +1,11 @@
+name: entrypoint-test
+version: latest
+base: ubuntu:20.04
+entrypoint: [/usr/bin/hello]
+cmd: [-g, "ship it!"]
```

Comment:
is this really used like this? Shouldn't entrypoint be `[/usr/bin/hello,  -g, "ship it!"]`

Reply:
Entrypoint and cmd are a bit confusing, but the idea is that if you define both values then `cmd` act as a complement to `entrypoint`. You can define only one of them, and if `entrypoint` is not defined the first `cmd` entry is the command to execute. `cmd` is also replaced by the command parameter if it's defined in the command-line.

Reply:
More information here: https://phoenixnap.com/kb/docker-cmd-vs-entrypoint

Reply:
I wonder if we should get rid of the entrypoint and cmd commands and do something like we do in snaps where a `daemon: ` converts this to a service. In any case we have not solidified the spec on this.

Reply:
The current entries map directly to the OCI spec and the concept should feel natural to dockerfile authors. We can change that to a different format, but having something that maps to `docker run --entrypoint` and a different thing that can be overriden using `docker run <command>` will probably be convenient/necessary.

Reply:
This is a good question to raise. Maybe I can shed some extra light on this.

### Meaning

The simplest way to look at it is: 
 - ENTRYPOINT is the image's executable. And even though it can be overridden at runtime, it shows what binary your container is going to run as PID 1 at startup
 - CMD provides default arguments to entrypoint

ENTRYPOINT always comes first, and depending on how they are set, the end result can behave very differently. This matrix summarizes their possible combinations: https://docs.docker.com/engine/reference/builder/#understand-how-cmd-and-entrypoint-interact

### Rule of thumb + examples

 1. "your image runs an executable which is not supposed to accept any arguments, i.e. it does one thing and one things only" -> `cmd: ["/usr/bin/hello", "-g", "ship it!"]`
 2. "your image runs an executable that by default prints "ship it!", but at runtime, users can print whatever they like" -> `entrypoint: ["/usr/bin/hello"]` + `cmd: ["-g", "ship it!"]`. Then at runtime, a user could do `docker run entrypoint-test -g "cool stuff"`, and "ship it" would no longer be printed
 3. "your image runs an executable which needs arguments at runtime, since its execution highly depends on the user configuration and underlying environment" -> `entrypoint: ["/usr/bin/hello"]` (no `cmd`)
 4. "your image runs an executable that always does one thing by default. It has one function. But this function can be extended with additional parameters if the users wants to" -> `entrypoint: ["/usr/bin/ls", "-l"]` (no `cmd`). Then at runtime, users could do something like `docker run entrypoint-test -h -a`
 5. "your image doesn't really have a default executable nor command..." -> not possible. At startup, the container runtime must at least have either an entrypoint or a default command to execute. This is the typical case of OS images like `ubuntu`. A common practice here is to set `cmd: ["bash"]`


 ### Formats

I personally don't like the way this was designed, but unfortunately, the current implementation of ENTRYPOINT and CMD has some entropy when it comes to how they are defined at build time (as you can see from the matrix at https://docs.docker.com/engine/reference/builder/#understand-how-cmd-and-entrypoint-interact). 

We either keep the same syntax and business logic as Dockerfiles (i.e. implicitly infer whether the user wants "exec form" or "shell form" based on whether the values are provided as a List[str] or str, respectively), **or**, we can try simplify this whilst providing the same functionality. I see 2 possible ways:
 1. **explicit declaration**: similar to Dockerfiles, but without any hidden assumptions. It would look something like this in the Dockerfile: 
 ```
 entrypoint:
    value: /usr/bin/hello
    format: exec  
 ```
 
2. **no format**: we'd let users literally write their ENTRYPOINT and CMD as they wish. So for example, if a user wants an entrypoint "/usr/bin/hello" to be executed in a shell, then he/she would have to write it like this in the rockcraft.yaml: `entrypoint: /bin/sh -c '/usr/bin/hello'`

Reply:
Wouldn't this clean things up a bit from an experience point of view?
```yaml
run:  # optional
    type: [entrypoint|cmd]  # defaults to entrypoint
    command: [<cmd>...]
    args: [<args>...]  # optional
```

I used `run` to avoid use of `command` or `entrypoint` as the main keyword to avoid confusion of intent.